i18n-helpers/src/bin/mdbook-i18n.rs

```diff
@@ -0,0 +1,107 @@
+    /// ```
+    #[serde(default)]
+    languages: BTreeMap<String, String>,
+    /// Default language code. It will not be translated.
+    default_language: Option<String>,
+
+    /// Whether to translate all languages or just the selected language, defaults to false.
+    #[serde(default)]
+    translate_all_languages: bool,
+    /// Whether to move the translations to the html directory, defaults to false.
+    ///
+    /// By default, translations' output will live in `book/i18n/<language>/<renderer>`.
+    /// For all renderers in this list, we will move individual translations to `book/<renderer>/<language>`.
+    #[serde(default)]
+    move_translations_directories: Vec<String>,
+}
+
+fn main() {
+    let mut stdin = io::stdin();
+
+    // Get the configs
+    let ctx = RenderContext::from_json(&mut stdin).unwrap();
+    let i18n_config: I18nConfiguration = ctx
+        .config
+        .get_deserialized_opt("output.i18n")
+        .unwrap()
+        .unwrap();
+
+    if !i18n_config.translate_all_languages {
+        return;
+    }
+
+    let mut mdbook = mdbook::MDBook::load(&ctx.root).expect("Failed to load book");
+    // Overwrite with current values from stdin. This is necessary because mdbook will add data to the config.
+    mdbook.book = ctx.book.clone();
+    mdbook.config = ctx.config.clone();
+    mdbook.root = ctx.root.clone();
+
+    let book_config = mdbook
+        .config
+        .get_mut("output.i18n")
+        .expect("No output.i18n config in book.toml");
+    // Set translate_all_languages to false for nested builds to prevent infinite recursion.
+    book_config
+        .as_table_mut()
+        .expect("output.i18n config in book.toml is not a table")
+        .insert(String::from("translate_all_languages"), false.into());
+
+    let output_directory = ctx.destination;
+    let default_language = &i18n_config.default_language;
+
+    for language in i18n_config.languages.keys() {
+        // Skip current language and default language.
+        if Some(language) == ctx.config.book.language.as_ref() {
+            continue;
+        }
+        if let Some(default_language) = default_language {
+            if default_language == language {
+                continue;
+            }
+        }
+        let translation_path = output_directory.join(language);
+
+        // Book doesn't implement clone, so we just mutate in place.
+        mdbook.config.book.language = Some(language.clone());
+        mdbook.config.book.multilingual = true;
+        mdbook.config.build.build_dir = translation_path;
+        mdbook
+            .build()
+            .unwrap_or_else(|_| panic!("Failed to build translation for language: {}", language));
+        for renderer in &i18n_config.move_translations_directories {
+            std::fs::create_dir_all(output_directory.parent().unwrap().join(renderer))
+                .unwrap_or_else(|_| panic!("Failed to create html directory in output directory"));
```

Comment:
The error message if off and this seems to be the place where you violate the contact of a mdbook renderer?

This also seems to be where you depend on the `html` renderer running before this one?

My guess is that there can be bugs hiding here: such as accidentally copying the _previous_ output of `mdbook build`?

Reply:
I forgot to replace the message when I changed the code here

Reply:
> This also seems to be where you depend on the `html` renderer running before this one?
> 
> My guess is that there can be bugs hiding here: such as accidentally copying the _previous_ output of `mdbook build`?

The questions here are the important part: we have to avoid getting ourselves into a TeX-like situation where you need to run `mdbook build` twice to get the correct final output :smile: